<a href="https://colab.research.google.com/github/joyc3lin/sqlite_database_operations/blob/main/HHA504_week3assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
#loading in packages
import pandas as pd
from sqlalchemy import create_engine
import sqlite3
import re
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import pearsonr

# **1. Data Exploration and Analysis**

In [4]:
# importing pricing transparency datasets
df_msmc = pd.read_excel('datasets/MountSinaiMedicalCenter_standardcharges.xlsx')
df_nyp = pd.read_json('datasets/NewYorkPresbyterianHospital_standardcharges.json')

In [5]:
print(df_msmc.shape)
df_msmc

(3577, 56)


Charge Description  \
0                                  HC BRIDGE ASSESSMENT   
1                  HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
2     HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   
3                          HC ELECTROCONVULSIVE THERAPY   
4                        HC PSYCH TREATMENT & EDUCATION   
...                                                 ...   
3572  Extensive O.R. procedure unrelated to principa...   
3573  Extensive O.R. procedure unrelated to principa...   
3574  Extensive O.R. procedure unrelated to principa...   
3575  Non-extensive O.R. proc unrelated to principal...   
3576  Non-extensive O.R. proc unrelated to principal...   

                                       Long Description Billing Code  \
0                   Psychiatric diagnostic evaluation\n        90791   
1                   Psychiatric diagnostic evaluation\n        90791   
2     Psychiatric diagnostic evaluation with medical...        90792   
3     Electroconvulsive therapy (includes necessary ...        90870   
4     Training and educational services related to t...        G0177   
...                                                 ...          ...   
3572  Extensive O.R. procedure unrelated to principa...          981   
3573  Extensive O.R. procedure unrelated to principa...          982   
3574  Extensive O.R. procedure unrelated to principa...          983   
3575  Non-extensive O.R. proc unrelated to principal...          987   
3576  Non-extensive O.R. proc unrelated to principal...          988   

     Revenue Code / DRG Type   Gross Charge     Cash Price  \
0                        513     156.000000            175   
1                        900     327.600000            315   
2                        900     327.600000            315   
3                        901    1154.400000          577.2   
4                        942     166.400000           83.2   
...                      ...            ...            ...   
3572                     MS   176654.394366  110622.909091   
3573                     MS   105720.700000   33313.777778   
3574                     MS    45953.375000       23374.75   
3575                     MS   196679.153846        44923.2   
3576                     MS    63918.578947      24418.625   

      AETNA BETTER HEALTH (MEDICAID)  AETNA COMMERCIAL  AETNA INTERNATIONAL  \
0                             10.920         46.800000           132.600000   
1                             22.932         98.280000           278.460000   
2                             22.932         98.280000           278.460000   
3                             80.808        346.320000           981.240000   
4                             11.648         49.920000           141.440000   
...                              ...               ...                  ...   
3572                             NaN      52996.318310        150156.235211   
3573                             NaN      31716.210000         89862.595000   
3574                             NaN      13786.012500         39060.368750   
3575                             NaN      59003.746154        167177.280769   
3576                             NaN      19175.573684         54330.792105   

      AETNA MEDICARE  ...  UNITED HEALTHCARE  \
0          21.840000  ...          32.760000   
1          45.864000  ...          68.796000   
2          45.864000  ...          68.796000   
3         161.616000  ...         242.424000   
4          23.296000  ...          34.944000   
...              ...  ...                ...   
3572    24731.615211  ...       37097.422817   
3573    14800.898000  ...       22201.347000   
3574     6433.472500  ...        9650.208750   
3575    27535.081538  ...       41302.622308   
3576     8948.601053  ...       13422.901579   

      UNITED HEALTHCARE MEDICARE / MEDICA / PREFERRED CARE PARTNERS  \
0                                             20.280000               
1                                             42.5

In [6]:
print(df_nyp.shape)
df_nyp

(6193, 47)


Code (CPT/DRG)                                        Description  \
0             96360        HC IV INFUSION HYDRATION INITIAL 31 MIN-1HR   
1             96361            HC IV INFUSION HYDRATION FLUIDS ADDL HR   
2             96365  HC IV INFUSION FOR THER/PROPH/DIAG INITIAL UP ...   
3             96367  HC IV INF THER/PROPH/DIAG ADDL SEQ NEW DRUG UP...   
4             96368      HC IV INFUSION FOR THER/PROPH/DIAG CONCURRENT   
...             ...                                                ...   
6188            982  EXTENSIVE O.R. PROCEDURES UNRELATED TO PRINCIP...   
6189            983  EXTENSIVE O.R. PROCEDURES UNRELATED TO PRINCIP...   
6190            987  NON-EXTENSIVE O.R. PROCEDURES UNRELATED TO PRI...   
6191            988  NON-EXTENSIVE O.R. PROCEDURES UNRELATED TO PRI...   
6192            989  NON-EXTENSIVE O.R. PROCEDURES UNRELATED TO PRI...   

      Rev Code  Inpatient/Outpatient  Gross Charges  Discounted Cash Price  \
0         0260  Inpatient/Outpatient     866.000000             866.000000   
1         0260  Inpatient/Outpatient     358.000000             358.000000   
2         0260  Inpatient/Outpatient    1397.000000            1397.000000   
3         0260  Inpatient/Outpatient     480.000000             480.000000   
4         0260  Inpatient/Outpatient     480.000000             480.000000   
...        ...                   ...            ...                    ...   
6188  Multiple             Inpatient  184095.129020          184095.129020   
6189  Multiple             Inpatient   85385.941857           85385.941857   
6190  Multiple             Inpatient  411665.999955          411665.999955   
6191  Multiple             Inpatient  126522.871140          126522.871140   
6192  Multiple             Inpatient   77081.639496           77081.639496   

            Aetna          Cigna Empire Blue Cross Blue Shield Emblem Health  \
0         641.706        494.486                        640.84      1388.424   
1         265.278        204.418                        264.92      1305.112   
2        1035.177        797.687                       1033.78      1475.508   
3          355.68         274.08                         355.2       1325.12   
4          355.68         274.08                         355.2       1325.12   
...           ...            ...                           ...           ...   
6188   86723.5232   98911.694706                      82299.84  115129.89148   
6189   57748.8352   65864.888166                  63931.455093   76664.51828   
6190  114736.9984  130862.199072                 214294.876425  152319.20576   
6191   58654.7264   66898.093812                    91907.6928   77867.13496   
6192   38116.5824   43473.507792                   44463.77208   50601.70336   

      ... Consumer Health Network         Devon        Equian  First Health  \
0     ...                393.5104      393.5104      393.5104      393.5104   
1     ...                162.6752      162.6752      162.6752      162.6752   
2     ...                634.7968      634.7968      634.7968      634.7968   
3     ...                 218.112       218.112       218.112       218.112   
4     ...                 218.112       218.112       218.112       218.112   
...   ...                     ...           ...           ...           ...   
6188  ...            83652.826627  83652.826627  83652.826627  83652.826627   
6189  ...             38799.37198   38799.37198   38799.37198   38799.37198   
6190  ...            187061.03038  187061.03038  187061.03038  187061.03038   
6191  ...            57491.992646  57491.992646  57491.992646  57491.992646   
6192  ...            35025.896987  35025.896987  35025.896987  35025.896987   

         Magnacare Multiplan/Beechstreet/PHCS           QHM     Worldwide  \
0         393.5104                   393.5104      393.5104      393.5104   
1         162.6752                   162.6752      162.6752      162.6752   
2         634.7968                   6

## Data Cleaning

### *Cleaning df_nyp*

In [7]:
# checking data types
for c in df_nyp.columns:
  print(f"{c}: {type(df_nyp[c][0])}")

Code (CPT/DRG): <class 'int'>
Description: <class 'str'>
Rev Code: <class 'str'>
Inpatient/Outpatient: <class 'str'>
Gross Charges: <class 'numpy.float64'>
Discounted Cash Price: <class 'numpy.float64'>
Aetna: <class 'float'>
Cigna: <class 'float'>
Empire Blue Cross Blue Shield: <class 'float'>
Emblem Health: <class 'float'>
United Health Group: <class 'float'>
Aetna Medicare: <class 'float'>
AgeWell Medicare: <class 'float'>
Emblem Medicare: <class 'float'>
Empire Medicare: <class 'float'>
Fidelis Medicare: <class 'float'>
Healthfirst Medicare: <class 'float'>
UHC Community Plan/United Medicare: <class 'float'>
VNS Medicare: <class 'float'>
WellCare Medicare: <class 'float'>
1199: <class 'int'>
Affinity Molina Essential: <class 'float'>
Affinity Molina Medicaid/CHP: <class 'float'>
Amida Care Medicaid: <class 'float'>
Emblem Medicaid/CHP: <class 'float'>
Empire Healthplus Essential: <class 'float'>
Empire Healthplus Exchange: <class 'float'>
Empire Healthplus Medicaid/CHP: <class 'flo

In [8]:
# looking for missing values
missing_values = df_nyp.isnull().sum()
missing_values

Code (CPT/DRG)                          0
Description                             0
Rev Code                                0
Inpatient/Outpatient                    0
Gross Charges                           0
Discounted Cash Price                   0
Aetna                                   0
Cigna                                   0
Empire Blue Cross Blue Shield           0
Emblem Health                           0
United Health Group                     0
Aetna Medicare                          0
AgeWell Medicare                        0
Emblem Medicare                         0
Empire Medicare                         0
Fidelis Medicare                        0
Healthfirst Medicare                    0
UHC Community Plan/United Medicare      0
VNS Medicare                            0
WellCare Medicare                       0
1199                                    0
Affinity Molina Essential               0
Affinity Molina Medicaid/CHP            0
Amida Care Medicaid               

In [9]:
# dropping row with missing data
df_nyp.dropna(inplace=True)
df_nyp.shape

(5912, 47)

In [10]:
#removing duplicate rows and columns
df_nyp.drop_duplicates(inplace=True)
df_nyp.loc[:,~df_nyp.T.duplicated(keep='first')]
print(df_nyp.shape)

(5912, 47)


In [11]:
# cleaning column names

# Function to remove white space and special characters from a value
def clean_value(value):
    cleaned_value = re.sub(r'\s+|[^a-zA-Z0-9]', '', str(value))
    return cleaned_value

df_nyp.rename(mapper=clean_value, axis=1, inplace=True)
df_nyp.sample(10)

CodeCPTDRG                                        Description RevCode  \
2576      46288                                REPAIR ANAL FISTULA    0360   
1077      74415                   HC X-RAY IVP WITH NEPHROTOMOGRAM    0320   
1138      77074                    HC X-RAY OSSEOUS SURVEY LIMITED    0320   
477       86638        HC LAB COXIELLA BURNETII (Q FEVER) ANTIBODY    0302   
3936      51798  HC MEASURE POST-VOIDING RESIDUAL URINE&/BLADDE...    0510   
3468      94664      HC RESPIRATORY THERAPY EQUIPMENT AEROSOL MASK    0410   
992       72072                    HC X-RAY THORACIC SPINE 3 VIEWS    0320   
2604      47490                            INCISION OF GALLBLADDER    0360   
4112      D0340                    HC DENTAL CEPHALOMETRIC FILM 2D    0512   
1388      70486                   HC CT MAXILLOFACIAL W/O CONTRAST    0351   

       InpatientOutpatient  GrossCharges  DiscountedCashPrice  \
2576  Inpatient/Outpatient       22862.0              22862.0   
1077  Inpatient/Outpatient        1642.0               1642.0   
1138  Inpatient/Outpatient         969.0                969.0   
477   Inpatient/Outpatient          48.0                 48.0   
3936  Inpatient/Outpatient         428.0                428.0   
3468  Inpatient/Outpatient         507.0                507.0   
992   Inpatient/Outpatient         688.0                688.0   
2604  Inpatient/Outpatient       23200.0              23200.0   
4112  Inpatient/Outpatient         486.0                486.0   
1388  Inpatient/Outpatient        2261.0               2261.0   

                       Aetna                   Cigna  \
2576                    9144                    7496   
1077                1216.722                 696.208   
1138                 718.029                 410.856   
477                    8.484                  10.199   
3936  Not separately payable  Not separately payable   
3468                 375.687                 289.497   
992                  509.808                 291.712   
2604                    9144                    7496   
4112  Not separately payable  Not separately payable   
1388                1675.401                 958.664   

     EmpireBlueCrossBlueShield            EmblemHealth  ...  \
2576                     11924                  9474.4  ...   
1077                 1344.0823              709.353668  ...   
1138                   431.893              419.724576  ...   
477                      8.484                   8.484  ...   
3936                     299.6  Not separately payable  ...   
3468                  177.7105                 407.628  ...   
992                   282.3381              373.640576  ...   
2604                     12938                  9474.4  ...   
4112                     340.2  Not separately payable  ...   
1388                 1011.3422              631.612576  ...   

     ConsumerHealthNetwork       Devon      Equian FirstHealth   Magnacare  \
2576            10388.4928  10388.4928  10388.4928  10388.4928  10388.4928   
1077              746.1248    746.1248    746.1248    746.1248    746.1248   
1138              440.3136    440.3136    440.3136    440.3136    440.3136   
477                21.8112     21.8112     21.8112     21.8112     21.8112   
3936              194.4832    194.4832    194.4832    194.4832    194.4832   
3468              230.3808    230.3808    230.3808    230.3808    230.3808   
992               312.6272    312.6272    312.6272    312.6272    312.6272   
2604              10542.08    10542.08    10542.08    10542.08    10542.08   
4112              220.8384    220.8384    220.8384    220.8384    220.8384   
1388             1027.3984   1027.3984   1027.3984   1027.3984   1027.3984   

     MultiplanBeechstreetPHCS         QHM   Worldwide MinimumNegotiatedCharge  \
2576               10388.4928  10388.4928  10388.4928             1769.500996   
1077                 746.1248    746.1248    746.1248              220.916500   
1138             

### Cleaing df_msmc

In [12]:
# checking data types
for c in df_msmc.columns:
  print(f"{c}: {type(df_msmc[c][0])}")

Charge Description: <class 'str'>
Long Description: <class 'str'>
Billing Code: <class 'int'>
Revenue Code / DRG Type: <class 'str'>
Gross Charge: <class 'numpy.float64'>
Cash Price: <class 'int'>
AETNA BETTER HEALTH (MEDICAID): <class 'numpy.float64'>
AETNA COMMERCIAL: <class 'numpy.float64'>
AETNA INTERNATIONAL: <class 'numpy.float64'>
AETNA MEDICARE: <class 'numpy.float64'>
AMBETTER: <class 'numpy.float64'>
AMERIGROUP: <class 'numpy.float64'>
BEECH STREET: <class 'numpy.float64'>
BCBS HEALTH OPTIONS HMO: <class 'numpy.float64'>
BCBS MEDICARE PPO: <class 'numpy.float64'>
BCBS HEALTH OPTIONS MCR: <class 'numpy.float64'>
BCBS NETWORK BLUE: <class 'numpy.float64'>
BCBS PPO: <class 'numpy.float64'>
BCBS PHS (TRADITIONAL): <class 'numpy.float64'>
BRIGHT HEALTH: <class 'numpy.float64'>
BRIGHT HEALTH MEDICARE: <class 'numpy.float64'>
CAREPLUS MEDICARE: <class 'numpy.float64'>
CIGNA : <class 'numpy.float64'>
CORVEL: <class 'numpy.float64'>
COVENTRY (FIRST HEALTH): <class 'numpy.float64'>
COV

In [13]:
# looking for missing values
missing_values2 = df_msmc.isnull().sum()
print(missing_values2)

Charge Description                                                 0
Long Description                                                   0
Billing Code                                                     729
Revenue Code / DRG Type                                            0
Gross Charge                                                       0
Cash Price                                                         0
AETNA BETTER HEALTH (MEDICAID)                                   572
AETNA COMMERCIAL                                                 270
AETNA INTERNATIONAL                                              270
AETNA MEDICARE                                                   270
AMBETTER                                                         270
AMERIGROUP                                                         0
BEECH STREET                                                     270
BCBS HEALTH OPTIONS HMO                                          270
BCBS MEDICARE PPO                 

In [14]:
# dropping row with missing data
df_msmc.dropna(inplace=True)
df_msmc.shape

(2006, 56)

In [15]:
#removing duplicate rows and columns
df_msmc.drop_duplicates(inplace=True)
df_msmc.loc[:,~df_msmc.T.duplicated(keep='first')]
print(df_msmc.shape)

(2003, 56)


In [16]:
# cleaning column names
df_msmc.rename(mapper=clean_value, axis=1, inplace=True)
df_msmc.sample(10)

ChargeDescription  \
977                          HC HIV GENOTYPE-   
2506         HC MR ANGIO PELVIS W/WO CONTRAST   
138   HC HUNTINGTON DISEASE MUTATION ANALYSIS   
832            HC NMDA RECEPTOR IGG CSF TITER   
509            HC INSULIN LIKE GROWTH FACT.II   
764                     HC BETA 2 TRANSFERRIN   
831                  HC CHIKUNGUNYA IGM TITER   
2269              HC PT PHYS PERF UE W/REPORT   
2262                  HC THERAPEUTIC EXERCISE   
1264                     HC XR SKULL COMPLETE   

                                        LongDescription BillingCode  \
977   Infectious agent genotype analysis by nucleic ...       87901   
2506  Magnetic resonance angiography without contras...       C8920   
138   HTT (huntingtin) (eg, Huntington disease) gene...       81271   
832   Fluorescent noninfectious agent antibody; tite...       86256   
509   Mass spectrometry and tandem mass spectrometry...       83789   
764   Immunofixation electrophoresis; other fluids w...       86335   
831            Antibody; virus, not elsewhere specified       86790   
2269  Physical performance test or measurement (eg, ...       97750   
2262  Therapeutic procedure, 1 or more areas, each 1...       97110   
1264  Radiologic examination, skull; complete, minim...       70260   

     RevenueCodeDRGType  GrossCharge CashPrice  AETNABETTERHEALTHMEDICAID  \
977                 306       156.00       150                    10.9200   
2506                610      6296.00      3148                   440.7200   
138                 300       283.92    141.96                    19.8744   
832                 302       299.52       288                    20.9664   
509                 301       104.78    100.75                     7.3346   
764                 302       200.90    193.17                    14.0630   
831                 302        74.68     37.34                     5.2276   
2269                420       237.10    118.55                    16.5970   
2262                420       229.37   114.685                    16.0559   
1264                320      1464.61       200                   102.5227   

      AETNACOMMERCIAL  AETNAINTERNATIONAL  AETNAMEDICARE  ...  \
977            46.800            132.6000        21.8400  ...   
2506         1888.800           5351.6000       881.4400  ...   
138            85.176            241.3320        39.7488  ...   
832            89.856            254.5920        41.9328  ...   
509            31.434             89.0630        14.6692  ...   
764            60.270            170.7650        28.1260  ...   
831            22.404             63.4780        10.4552  ...   
2269           71.130            201.5350        33.1940  ...   
2262           68.811            194.9645        32.1118  ...   
1264          439.383           1244.9185       205.0454  ...   

      UNITEDHEALTHCARE  UNITEDHEALTHCAREMEDICAREMEDICAPREFERREDCAREPARTNERS  \
977            32.7600                                            20.2800     
2506         1322.1600                                           818.4800     
138            59.6232                                            36.9096     
832            62.8992                                            38.9376     
509            22.0038                                            13.6214     
764            42.1890                                            26.1170     
831            15.6828                                             9.7084     
2269           49.7910                                            30.8230     
2262           48.1677                                            29.8181     
1264          307.5681                                           190.3993     

      UNITEDHEALTHCAREOPTIONSPPO  UNITEDHEALTHCAREMEDICAID  WELLCAREMEDICAID  \
977                      34.3200                   10.9200           10.9200   
2506                   1385.1200                  440.7200          440.7200   
138                      62.4624

## data distribution and statistics

### *df_nyp*

In [17]:
# distribution by RevCode
df_nyp.groupby(['RevCode']).describe()

GrossCharges                                              \
                count           mean            std           min   
RevCode                                                             
0260             12.0     753.500000     404.300406    357.000000   
0270             10.0      67.200000      61.372813      2.000000   
0274              2.0    1003.500000     341.532575    762.000000   
0278             10.0    4298.600000   12854.565321     83.000000   
0300             26.0    3534.076923    7531.502899     36.000000   
...               ...            ...            ...           ...   
0942              8.0     170.250000     138.450352     65.000000   
0943              2.0     388.000000       9.899495    381.000000   
0944              1.0     131.000000            NaN    131.000000   
0948              2.0     197.000000       0.000000    197.000000   
Multiple        723.0  180613.709921  223844.512646  16861.788766   

                                                                   \
                  25%            50%            75%           max   
RevCode                                                             
0260        480.00000     556.000000     991.000000  1.397000e+03   
0270         26.25000      56.500000      85.000000  1.890000e+02   
0274        882.75000    1003.500000    1124.250000  1.245000e+03   
0278        116.50000     166.500000     364.250000  4.087900e+04   
0300        112.50000     535.000000    2752.750000  2.815100e+04   
...               ...            ...            ...           ...   
0942         87.75000      90.000000     219.250000  4.550000e+02   
0943        384.50000     388.000000     391.500000  3.950000e+02   
0944        131.00000     131.000000     131.000000  1.310000e+02   
0948        197.00000     197.000000     197.000000  1.970000e+02   
Multiple  73238.72433  114634.634066  194664.056085  2.538980e+06   

         DiscountedCashPrice                 ... MinimumNegotiatedCharge  \
                       count           mean  ...                     75%   
RevCode                                      ...                           
0260                    12.0     753.500000  ...              253.048250   
0270                    10.0      67.200000  ...               38.624000   
0274                     2.0    1003.500000  ...              337.500000   
0278                    10.0    4298.600000  ...              131.130000   
0300                    26.0    3534.076923  ...              332.605219   
...                      ...            ...  ...                     ...   
0942                     8.0     170.250000  ...               41.885000   
0943                     2.0     388.000000  ...               94.574850   
0944                     1.0     131.000000  ...               59.526400   
0948                     2.0     197.000000  ...               70.413000   
Multiple               723.0  180613.709921  ...            14066.153064   

                      MaximumNegotiatedCharge                                \
                  max                   count           mean            std   
RevCode                                                                       
0260         551.8625                    12.0    1482.603223     245.133340   
0270          85.8816                    10.0      54.028800      49.343742   
0274         358.5600                     2.0     455.990400     155.192402   
0278       11773.1520                    10.0    6212.815508   10714.042758   
0300         532.0000                    26.0    2793.977263    6051.885289   
...               ...                     ...            ...            ...   
0942         206.7520                     8.0     314.472284     120.306293   
0943         103.7530                     2.0     311.952000       7.959194   
0944          59.5264                     1.0     468.948267            NaN   
0948          70.4130                     2.0     182.981700  

In [18]:
# value_counts of description
desc_counts = df_nyp['Description'].value_counts()
print(desc_counts)


KNEE ARTHROSCOPY/SURGERY                                                         16
CYSTOSCOPY AND TREATMENT                                                         12
REPAIR EYELID DEFECT                                                             12
TREAT HUMERUS FRACTURE                                                            9
SHOULDER ARTHROSCOPY/SURGERY                                                      7
                                                                                 ..
PARTIAL REMOVAL LEG BONE(S)                                                       1
REMOVE FEMUR LESION/GRAFT                                                         1
REMOVE FEMUR LESION                                                               1
EXC THIGH/KNEE TUM DEP 5CM/>                                                      1
NON-EXTENSIVE O.R. PROCEDURES UNRELATED TO PRINCIPAL DIAGNOSIS WITHOUT CC/MCC     1
Name: Description, Length: 5652, dtype: int64


In [19]:
# value_counts of InpatientOutpatient
io_counts = df_nyp['InpatientOutpatient'].value_counts()
print(io_counts)

Inpatient/Outpatient    5189
Inpatient                723
Name: InpatientOutpatient, dtype: int64


In [20]:
# value_counts of CodeCPTDRG
code_counts = df_nyp['CodeCPTDRG'].value_counts()
print(code_counts)

96360    1
54056    1
57105    1
57065    1
56606    1
        ..
27827    1
27823    1
27822    1
27814    1
989      1
Name: CodeCPTDRG, Length: 5912, dtype: int64


In [21]:
# statistics of GrossCharges
print(f"The mean of GrossCharges is : {df_nyp['GrossCharges'].mean()}")
print(f"The median of GrossCharges is : {df_nyp['GrossCharges'].median()}")
print(f"The mode of GrossCharges is : {df_nyp['GrossCharges'].mode().iloc[0]}")
print(f"The range of GrossCharges is: {df_nyp['GrossCharges'].max() - df_nyp['GrossCharges'].min()}")
print(f"The variance of GrossCharges is: {df_nyp['GrossCharges'].var()}")
print(f"The standard deviation of GrossCharges is: {df_nyp['GrossCharges'].std()}")
print(f"The IQR of GrossCharges is: {df_nyp['GrossCharges'].quantile(0.75) - df_nyp['GrossCharges'].quantile(0.25)}")

The mean of GrossCharges is : 33550.80400420582
The median of GrossCharges is : 3968.0
The mode of GrossCharges is : 122.0
The range of GrossCharges is: 2538978.617631959
The variance of GrossCharges is: 9445426925.583761
The standard deviation of GrossCharges is: 97187.58627306145
The IQR of GrossCharges is: 31052.5


In [22]:
# statistics of DiscountedCashPrice
print(f"The mean of DiscountedCashPrice is : {df_nyp['DiscountedCashPrice'].mean()}")
print(f"The median of DiscountedCashPrice is : {df_nyp['DiscountedCashPrice'].median()}")
print(f"The mode of DiscountedCashPrice is : {df_nyp['DiscountedCashPrice'].mode().iloc[0]}")
print(f"The range of DiscountedCashPrice is: {df_nyp['DiscountedCashPrice'].max() - df_nyp['DiscountedCashPrice'].min()}")
print(f"The variance of DiscountedCashPrice is: {df_nyp['DiscountedCashPrice'].var()}")
print(f"The standard deviation of DiscountedCashPrice is: {df_nyp['DiscountedCashPrice'].std()}")
print(f"The IQR of DiscountedCashPrice is: {df_nyp['DiscountedCashPrice'].quantile(0.75) - df_nyp['DiscountedCashPrice'].quantile(0.25)}")

The mean of DiscountedCashPrice is : 33550.80400420582
The median of DiscountedCashPrice is : 3968.0
The mode of DiscountedCashPrice is : 122.0
The range of DiscountedCashPrice is: 2538978.617631959
The variance of DiscountedCashPrice is: 9445426925.583761
The standard deviation of DiscountedCashPrice is: 97187.58627306145
The IQR of DiscountedCashPrice is: 31052.5


In [23]:
# statistics of MinimumNegotiatedCharge
print(f"The mean of MinimumNegotiatedCharge	is : {df_nyp['MinimumNegotiatedCharge'].mean()}")
print(f"The median of MinimumNegotiatedCharge	is : {df_nyp['MinimumNegotiatedCharge'].median()}")
print(f"The mode of MinimumNegotiatedCharge	is : {df_nyp['MinimumNegotiatedCharge'].mode().iloc[0]}")
print(f"The range of MinimumNegotiatedCharge is: {df_nyp['MinimumNegotiatedCharge'].max() - df_nyp['MinimumNegotiatedCharge'].min()}")
print(f"The variance of MinimumNegotiatedCharge is: {df_nyp['MinimumNegotiatedCharge'].var()}")
print(f"The standard deviation of MinimumNegotiatedCharge is: {df_nyp['MinimumNegotiatedCharge'].std()}")
print(f"The IQR of MinimumNegotiatedCharge is: {df_nyp['MinimumNegotiatedCharge'].quantile(0.75) - df_nyp['MinimumNegotiatedCharge'].quantile(0.25)}")

The mean of MinimumNegotiatedCharge	is : 2249.9188801499627
The median of MinimumNegotiatedCharge	is : 431.1429045953
The mode of MinimumNegotiatedCharge	is : 2412.0
The range of MinimumNegotiatedCharge is: 204860.6716
The variance of MinimumNegotiatedCharge is: 36188853.47582183
The standard deviation of MinimumNegotiatedCharge is: 6015.717203777272
The IQR of MinimumNegotiatedCharge is: 1984.7715625


In [24]:
# statistics of MaximumNegotiatedCharge
print(f"The mean of MaximumNegotiatedCharge	is : {df_nyp['MaximumNegotiatedCharge'].mean()}")
print(f"The median of MaximumNegotiatedCharge	is : {df_nyp['MaximumNegotiatedCharge'].median()}")
print(f"The mode of MaximumNegotiatedCharge	is : {df_nyp['MaximumNegotiatedCharge'].mode().iloc[0]}")
print(f"The range of MaximumNegotiatedCharge is: {df_nyp['MaximumNegotiatedCharge'].max() - df_nyp['MaximumNegotiatedCharge'].min()}")
print(f"The variance of MaximumNegotiatedCharge is: {df_nyp['MaximumNegotiatedCharge'].var()}")
print(f"The standard deviation of MaximumNegotiatedCharge is: {df_nyp['MaximumNegotiatedCharge'].std()}")
print(f"The IQR of MaximumNegotiatedCharge is: {df_nyp['MaximumNegotiatedCharge'].quantile(0.75) - df_nyp['MaximumNegotiatedCharge'].quantile(0.25)}")

The mean of MaximumNegotiatedCharge	is : 23902.77424282837
The median of MaximumNegotiatedCharge	is : 4086.7313
The mode of MaximumNegotiatedCharge	is : 11924.0
The range of MaximumNegotiatedCharge is: 1713458.5920000002
The variance of MaximumNegotiatedCharge is: 5092967278.384099
The standard deviation of MaximumNegotiatedCharge is: 71365.02839895812
The IQR of MaximumNegotiatedCharge is: 18618.398999999998


### *df_msmc*

In [25]:
# distribution of RevenueCodeDRGType

df_msmc.groupby(['RevenueCodeDRGType']).describe()

GrossCharge                                                \
                         count         mean           std     min        25%   
RevenueCodeDRGType                                                             
250                        2.0    10.400000      4.412346    7.28     8.8400   
260                       18.0   574.358889    315.333955  253.12   299.2500   
272                        4.0  3498.950000   6584.163770   61.30   101.8900   
278                       16.0  5795.913750  14727.762487   23.40    96.8500   
289                        1.0   624.000000           NaN  624.00   624.0000   
...                        ...          ...           ...     ...        ...   
920                        2.0   640.315000    524.934861  269.13   454.7225   
921                       19.0  1394.644211    422.736078  624.50  1038.3450   
940                        1.0   132.130000           NaN  132.13   132.1300   
942                        1.0   166.400000           NaN  166.40   166.4000   
943                        4.0   281.655000    156.389285   96.60   209.2650   

                                                  AETNABETTERHEALTHMEDICAID  \
                         50%        75%       max                     count   
RevenueCodeDRGType                                                            
250                   10.400    11.9600     13.52                       2.0   
260                  445.000   787.2200   1126.25                      18.0   
272                  281.310  3678.3700  13371.88                       4.0   
278                  161.720   728.5200  43494.10                      16.0   
289                  624.000   624.0000    624.00                       1.0   
...                      ...        ...       ...                       ...   
920                  640.315   825.9075   1011.50                       2.0   
921                 1279.860  1771.5000   1964.16                      19.0   
940                  132.130   132.1300    132.13                       1.0   
942                  166.400   166.4000    166.40                       1.0   
943                  277.770   350.1600    474.48                       4.0   

                                ... MaximumNegotiatedCharge              \
                          mean  ...                     75%         max   
RevenueCodeDRGType              ...                                       
250                   0.728000  ...               10.166000     11.4920   
260                  40.205122  ...              669.137000    957.3125   
272                 244.926500  ...             3126.614500  11366.0980   
278                 405.713963  ...              619.242000  36969.9850   
289                  43.680000  ...              530.400000    530.4000   
...                        ...  ...                     ...         ...   
920                  44.822050  ...              702.021375    859.7750   
921                  97.625095  ...             1505.775000   1669.5360   
940                   9.249100  ...              112.310500    112.3105   
942                  11.648000  ...              141.440000    141.4400   
943                  19.715850  ...              297.636000    403.3080   

                   MinimumNegotiatedCharge                                    \
                                     count        mean          std      min   
RevenueCodeDRGType                                                             
250                                    2.0    0.728000     0.308864   0.5096   
260                                   18.0   40.205122    22.073377  17.7184   
272                                    4.0  244.926500   460.891464   4.2910   
278                                   16.0  405.713963  1030.943374   1.6380   
289                                    1.0   43.680000          NaN  43.6800   
...                                    ...         ...          ...      ...   
920                 

In [26]:
# value_counts of ChargeDescription
cdesc_counts = df_msmc['ChargeDescription'].value_counts()
print(cdesc_counts)

HC PHENOBARBITAL                                                               2
HC MR BRAIN W/WO CONTRAST                                                      2
HC 3D RENDERING W POST PROCESS                                                 2
HC C3 NEPHRITIC FACTOR                                                         2
HC PLACE DEVICE/MARKER, NON PROS                                               2
                                                                              ..
HC ANTIMYELOPEROX AB QT                                                        1
HC ANTI-MITOCHONDRIA AB                                                        1
HC C1Q                                                                         1
HC ANTICARDIOLIPIN AB                                                          1
HC REM MNTR PHYSIOL PARAM 1ST DEV SUPPLY  WITH DAILY RECORDINGS; EA 30 DAYS    1
Name: ChargeDescription, Length: 1995, dtype: int64


In [27]:
# value_counts of BillingCode
bcode_counts = df_msmc['BillingCode'].value_counts()
print(bcode_counts)

G0463    13
86255    13
83520    12
94640    10
78830     9
         ..
88364     1
88365     1
88367     1
88368     1
99453     1
Name: BillingCode, Length: 1391, dtype: int64


In [28]:
# statistics of GrossCharge
print(f"The mean of GrossCharge is : {df_msmc['GrossCharge'].mean()}")
print(f"The median of GrossCharge is : {df_msmc['GrossCharge'].median()}")
print(f"The mode of GrossCharge is : {df_msmc['GrossCharge'].mode().iloc[0]}")
print(f"The range of GrossCharge is: {df_msmc['GrossCharge'].max() - df_msmc['GrossCharge'].min()}")
print(f"The variance of GrossCharge is: {df_msmc['GrossCharge'].var()}")
print(f"The standard deviation of GrossCharge is: {df_msmc['GrossCharge'].std()}")
print(f"The IQR of GrossCharge is: {df_msmc['GrossCharge'].quantile(0.75) - df_msmc['GrossCharge'].quantile(0.25)}")

The mean of GrossCharge is : 1753.8667099350973
The median of GrossCharge is : 253.12
The mode of GrossCharge is : 12.48
The range of GrossCharge is: 62035.14
The variance of GrossCharge is: 29649627.837086324
The standard deviation of GrossCharge is: 5445.147182316225
The IQR of GrossCharge is: 987.505


In [29]:
# statistics of CashPrice
print(f"The mean of CashPrice is : {df_msmc['CashPrice'].mean()}")
print(f"The median of CashPrice is : {df_msmc['CashPrice'].median()}")
print(f"The mode of CashPrice is : {df_msmc['CashPrice'].mode().iloc[0]}")
print(f"The range of CashPrice is: {df_msmc['CashPrice'].max() - df_msmc['CashPrice'].min()}")
print(f"The variance of CashPrice is: {df_msmc['CashPrice'].var()}")
print(f"The standard deviation of CashPrice is: {df_msmc['CashPrice'].std()}")
print(f"The IQR of CashPrice is: {df_msmc['CashPrice'].quantile(0.75) - df_msmc['CashPrice'].quantile(0.25)}")

The mean of CashPrice is : 809.2674413379931
The median of CashPrice is : 109.2
The mode of CashPrice is : 125
The range of CashPrice is: 31017.57
The variance of CashPrice is: 7351213.61745605
The standard deviation of CashPrice is: 2711.312157877814
The IQR of CashPrice is: 422.305


In [30]:
# statistics of MinimumNegotiatedCharge
print(f"The mean of MinimumNegotiatedCharge	is : {df_msmc['MinimumNegotiatedCharge'].mean()}")
print(f"The median of MinimumNegotiatedCharge	is : {df_msmc['MinimumNegotiatedCharge'].median()}")
print(f"The mode of MinimumNegotiatedCharge	is : {df_msmc['MinimumNegotiatedCharge'].mode().iloc[0]}")
print(f"The range of MinimumNegotiatedCharge is: {df_msmc['MinimumNegotiatedCharge'].max() - df_msmc['MinimumNegotiatedCharge'].min()}")
print(f"The variance of MinimumNegotiatedCharge is: {df_msmc['MinimumNegotiatedCharge'].var()}")
print(f"The standard deviation of MinimumNegotiatedCharge is: {df_msmc['MinimumNegotiatedCharge'].std()}")
print(f"The IQR of MinimumNegotiatedCharge is: {df_msmc['MinimumNegotiatedCharge'].quantile(0.75) - df_msmc['MinimumNegotiatedCharge'].quantile(0.25)}")

The mean of MinimumNegotiatedCharge	is : 122.77066969545683
The median of MinimumNegotiatedCharge	is : 17.718400000000003
The mode of MinimumNegotiatedCharge	is : 0.8736000000000002
The range of MinimumNegotiatedCharge is: 4342.4598000000005
The variance of MinimumNegotiatedCharge is: 145283.17640172297
The standard deviation of MinimumNegotiatedCharge is: 381.1603027621357
The IQR of MinimumNegotiatedCharge is: 69.12535000000001


In [31]:
# statistics of MaximumNegotiatedCharge
print(f"The mean of MaximumNegotiatedCharge	is : {df_msmc['MaximumNegotiatedCharge'].mean()}")
print(f"The median of MaximumNegotiatedCharge	is : {df_msmc['MaximumNegotiatedCharge'].median()}")
print(f"The mode of MaximumNegotiatedCharge	is : {df_msmc['MaximumNegotiatedCharge'].mode().iloc[0]}")
print(f"The range of MaximumNegotiatedCharge is: {df_msmc['MaximumNegotiatedCharge'].max() - df_msmc['MaximumNegotiatedCharge'].min()}")
print(f"The variance of MaximumNegotiatedCharge is: {df_msmc['MaximumNegotiatedCharge'].var()}")
print(f"The standard deviation of MaximumNegotiatedCharge is: {df_msmc['MaximumNegotiatedCharge'].std()}")
print(f"The IQR of MaximumNegotiatedCharge is: {df_msmc['MaximumNegotiatedCharge'].quantile(0.75) - df_msmc['MaximumNegotiatedCharge'].quantile(0.25)}")

The mean of MaximumNegotiatedCharge	is : 1490.7867034448327
The median of MaximumNegotiatedCharge	is : 215.152
The mode of MaximumNegotiatedCharge	is : 10.608
The range of MaximumNegotiatedCharge is: 52729.869
The variance of MaximumNegotiatedCharge is: 21421856.112294868
The standard deviation of MaximumNegotiatedCharge is: 4628.375104968791
The IQR of MaximumNegotiatedCharge is: 839.37925


### **Conclusions:**  

NYP has no cash discount as the basic satistical analysis returned the same values for gross charges and the discounted cash prices. MSMC does offer discounts for cash prices as the values of their satistical analyses were less than those for the gross charges.

# **2. SQLite Database Operations**

## 2.1 Local SQLite database

In [32]:
# creating a local database
conn = sqlite3.connect('health.db')
c = conn.cursor()

## 2.2 Manual table creation

In [33]:
# creating table

c.execute("""
            CREATE TABLE icecream
              (
                icecream_name text,
                icecream_type text,
                flavor text,
                servings_per_box real,
                cost_per_box real
              );
        """)

conn.commit()

In [34]:
# inserting data into icecream

sql_query = """

INSERT INTO icecream (
  'icecream_name',
  'icecream_type',
  'flavor',
  'servings_per_box',
  'cost_per_box'
  )
  values (
    'Dazs-genhaa',
    'bars',
    'vanilla',
    8,
    10
  );


"""
print(sql_query)



INSERT INTO icecream (
  'icecream_name',
  'icecream_type',
  'flavor',
  'servings_per_box',
  'cost_per_box'
  )
  values (
    'Dazs-genhaa',
    'bars',
    'vanilla',
    8,
    10
  );





In [35]:
# executing sql_query
c.execute(sql_query)
conn.commit()

In [36]:
# inserting more information
sql_query2 = """

INSERT INTO icecream (
  'icecream_name',
  'icecream_type',
  'flavor',
  'servings_per_box',
  'cost_per_box'
  )
  values (
    'Jen & Berrys',
    'sandwich',
    'coffee',
    12,
    18
  );


"""
print(sql_query2)



INSERT INTO icecream (
  'icecream_name',
  'icecream_type',
  'flavor',
  'servings_per_box',
  'cost_per_box'
  )
  values (
    'Jen & Berrys',
    'sandwich',
    'coffee',
    12,
    18
  );





In [37]:
# executing sql_query2

c.execute(sql_query2)
conn.commit()

In [38]:
# checking if rows have been populated into table
query = """

  SELECT *
  from icecream;

  """

c.execute(query)
print(c.fetchall())

[('Dazs-genhaa', 'bars', 'vanilla', 8.0, 10.0), ('Jen & Berrys', 'sandwich', 'coffee', 12.0, 18.0)]


In [39]:
# creating engine to connect to db

engine = create_engine('sqlite:///health.db')

In [40]:
# displaying table
icecream = pd.read_sql(query, conn)
icecream

icecream_name icecream_type   flavor  servings_per_box  cost_per_box
0   Dazs-genhaa          bars  vanilla               8.0          10.0
1  Jen & Berrys      sandwich   coffee              12.0          18.0

## 2.3 Automatic table creation

In [41]:
#creating new column and modifying the column names in icecream
columnNames = list(df_msmc)
idVars = columnNames[:5]
valueVars = columnNames[5:]

icecream_modified = df_msmc.melt(id_vars=idVars, value_vars=valueVars)

icecream_modified.columns

icecream_modified.rename(columns={'icecream_name': 'ChargeDescription',
                                  'icecream_type': 'LongDescription',
                                  'flavor':'BillingCode',
                                  'servings_per_box':'RevenueCodeDRGType',
                                  'cost_per_box': 'GrossCharge'}, inplace=True)

icecream_modified

ChargeDescription  \
0                                    HC BRIDGE ASSESSMENT   
1                    HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
2       HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   
3                            HC ELECTROCONVULSIVE THERAPY   
4                          HC PSYCH TREATMENT & EDUCATION   
...                                                   ...   
102148                                      HC PHLEBOTOMY   
102149                           HC CARDIAC REHAB PHASE I   
102150                           HC CARDIAC REHAB PHASE 2   
102151  HC REM MNTR PHYSIOL PARAM INITIAL SETUP AND ED...   
102152  HC REM MNTR PHYSIOL PARAM 1ST DEV SUPPLY  WITH...   

                                          LongDescription BillingCode  \
0                     Psychiatric diagnostic evaluation\n       90791   
1                     Psychiatric diagnostic evaluation\n       90791   
2       Psychiatric diagnostic evaluation with medical...       90792   
3       Electroconvulsive therapy (includes necessary ...       90870   
4       Training and educational services related to t...       G0177   
...                                                   ...         ...   
102148       Phlebotomy, therapeutic (separate procedure)       99195   
102149  Physician or other qualified health care profe...       93797   
102150  Physician or other qualified health care profe...       93798   
102151  Remote monitoring of physiologic parameter(s) ...       99453   
102152  Remote monitoring of physiologic parameter(s) ...       99454   

       RevenueCodeDRGType  GrossCharge                 variable    value  
0                     513       156.00                CashPrice      175  
1                     900       327.60                CashPrice      315  
2                     900       327.60                CashPrice      315  
3                     901      1154.40                CashPrice    577.2  
4                     942       166.40                CashPrice     83.2  
...                   ...          ...                      ...      ...  
102148                940       132.13  MinimumNegotiatedCharge   9.2491  
102149                943       246.82  MinimumNegotiatedCharge  17.2774  
102150                943       474.48  MinimumNegotiatedCharge  33.2136  
102151                943       308.72  MinimumNegotiatedCharge  21.6104  
102152                943        96.60  MinimumNegotiatedCharge    6.762  

[102153 rows x 7 columns]

In [42]:
# replacing icecream with icecream_modified
icecream_modified.to_sql('icecream', conn, if_exists='replace')

102153

In [46]:
#testing modified table

modified_query = """
  select *
  from icecream
  where RevenueCodeDRGType = '900'
  limit 200;
"""
check = pd.read_sql(modified_query, conn)
check

index                                  ChargeDescription  \
0         1               HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
1         2  HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   
2      2004               HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
3      2005  HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   
4      4007               HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
..      ...                                                ...   
97    96146  HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   
98    98148               HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
99    98149  HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   
100  100151               HC PSYCHIATRIC DIAGNOSTIC EVALUATION   
101  100152  HC PSYCHIATRIC DIAGNOSTIC EVALUATION W/ MEDICA...   

                                       LongDescription BillingCode  \
0                  Psychiatric diagnostic evaluation\n       90791   
1    Psychiatric diagnostic evaluation with medical...       90792   
2                  Psychiatric diagnostic evaluation\n       90791   
3    Psychiatric diagnostic evaluation with medical...       90792   
4                  Psychiatric diagnostic evaluation\n       90791   
..                                                 ...         ...   
97   Psychiatric diagnostic evaluation with medical...       90792   
98                 Psychiatric diagnostic evaluation\n       90791   
99   Psychiatric diagnostic evaluation with medical...       90792   
100                Psychiatric diagnostic evaluation\n       90791   
101  Psychiatric diagnostic evaluation with medical...       90792   

    RevenueCodeDRGType  GrossCharge                   variable   value  
0                  900        327.6                  CashPrice     315  
1                  900        327.6                  CashPrice     315  
2                  900        327.6  AETNABETTERHEALTHMEDICAID  22.932  
3                  900        327.6  AETNABETTERHEALTHMEDICAID  22.932  
4                  900        327.6            AETNACOMMERCIAL   98.28  
..                 ...          ...                        ...     ...  
97                 900        327.6                WWCONCIERGE   163.8  
98                 900        327.6    MaximumNegotiatedCharge  278.46  
99                 900        327.6    MaximumNegotiatedCharge  278.46  
100                900        327.6    MinimumNegotiatedCharge  22.932  
101                900        327.6    MinimumNegotiatedCharge  22.932  

[102 rows x 8 columns]

In [47]:
# making another query

modified_query2 = """
  select GrossCharge
  from icecream
  where BillingCode = '90791'
  limit 200;
"""
check2 = pd.read_sql(modified_query2, conn)
check2

GrossCharge
0          156.0
1          327.6
2          156.0
3          327.6
4          156.0
..           ...
97         327.6
98         156.0
99         327.6
100        156.0
101        327.6

[102 rows x 1 columns]

### **analysis:**

Modified_query: Ran select query to only show rows from the icecream table for a specific revenue code number: 900. Additionally, the select only asked for the first 200 results to be shown. The results returned only 102 rows because there was only 102 rows with code 900.

Modified_query2: ran select query to only take from gross charge column of icecream table. also asked for gross charges with billing code of 90781. The limit was also set to 200 and only 102 results were returned. The results only showed the grosscharges and not the billing code even though they were chosen based on their billing code number.